granary/flickr.py

```diff
@@ -58,6 +62,161 @@ def call_api_method(self, method, params={}):
     return flickr_auth.call_api_method(
       method, params, self.access_token_key, self.access_token_secret)
 
+  def upload_photo(self, params, photo_file):
+    """Upload a photo via the Flickr API
+    """
+    return flickr_auth.upload(
+      params, photo_file, self.access_token_key, self.access_token_secret)
+
+  def create(self, obj, include_link=False):
+    """Creates a photo, comment, or favorite.
+
+    Args:
+      obj: ActivityStreams object
+      include_link: boolean
+
+    Returns:
+      a CreationResult whose content will be a dict with 'id', 'url',
+      and 'type' keys (all optional) for the newly created Flickr
+      object (or None)
+
+    """
+    return self._create(obj, preview=False, include_link=include_link)
+
+  def preview_create(self, obj, include_link=False):
+    """Preview creation of a photo, comment, or favorite.
+
+    Args:
+      obj: ActivityStreams object
+      include_link: boolean
+
+    Returns:
+      a CreationResult whose description will be an HTML summary of
+      what publishing will do, and whose content will be an HTML preview
+      of the result (or None)
+    """
+    return self._create(obj, preview=True, include_link=include_link)
+
+  def _create(self, obj, preview, include_link):
+    """Creates or previews creating for the previous two methods.
+
+    https://www.flickr.com/services/api/upload.api.html
+    https://www.flickr.com/services/api/flickr.photos.comments.addComment.html
+    https://www.flickr.com/services/api/flickr.favorites.add.html
+
+    Args:
+      obj: ActivityStreams object
+      preview: boolean
+      include_link: boolean
+
+    Return:
+      a CreationResult
+    """
+    # photo, comment, or like
+    type = source.object_type(obj)
+    logging.debug('publishing object type %s to Flickr', type)
+    content = self._content_for_create(obj)
+
+    include_url = ('(<a href="%s">Original</a>)' % obj.get('url')
+                   if include_link else None)
+
+    if obj.get('image') and type in ('note', 'article'):
+      image_url = obj.get('image').get('url')
+      title = obj.get('displayName')
+
+      if title and content and content.strip() == title.strip():
```

Comment:
Well at least I should use `mf2util.is_name_a_title` which is a little more robust and more self-documenting